ath79: initial support for Ubiquiti RouterStation and RouterStation Pro

tofurky opened the pull request:
This adds a shared ar7161_ubnt_rs.dtsi as well as two other .dts files
that utilize it, ar7161_ubnt_rs.dts and ar7161_ubnt_rspro.dts.

The modifications to generic-ubnt.mk, config-default, and base-files
necessary for image generation, parsing RedBoot FIS partitions, network
configuration, and sysupgrade are also included. This reintroduces
platform_do_upgrade_combined() and its supporting functions to
/lib/upgrade/platform.sh, which were previously removed from ath79 in
3e9d9f62258f80298710441e0db557e59e152dcf "ath79: sysupgrade: drop unused
platform checks".

All hardware is functional including the AR8216 switch (for the Pro),
wireless via ath5k/ath9k using the miniPCI slots, flash, USB, button,
and LED.

Switch and LAN/WAN configuration is the same as it is with the
equivalent ar71xx targets. MAC addresses are assigned based upon the
content stored in the RedBoot config partition.

Flashing via both sysupgrade and TFTP has been confirmed to work. Also,
the initramfs images are now raw .bin files instead of being wrapped in
a uImage (as they currently are in ar71xx), which makes them bootable
with RedBoot.

One noteable difference to ar71xx is the inclusion of the RedBoot
"fconfig" utility in DEVICE_PACKAGES. The justification for this is that
many other boards already include the corresponding U-Boot utilities
fw_printenv and fw_setenv.

Signed-off-by: Tomasz Maciej Nowak <tomek_n@o2.pl>
Signed-off-by: Matt Merhar <mattmerhar@protonmail.com>

mkresin commented on target/linux/ath79/base-files/lib/upgrade/platform.sh:
```diff
@@ -5,14 +5,112 @@
+	PLATFORM_DO_UPGRADE_COMBINED_SEPARATE_MTD=0
+}
+
+platform_check_image() {
```
due to the image metadata, such checks are no longer required. drop it.

tmn505 replied:
Agreed.

mkresin commented on target/linux/ath79/base-files/lib/upgrade/platform.sh:
```diff
@@ -5,14 +5,112 @@
+	done
+}
+
+platform_check_combined() {
```
I guess it is time to get rid of this check as well. We should trust that we generate the correct image format.

tmn505 replied:
Agreed.

mkresin commented on target/linux/ath79/base-files/lib/upgrade/platform.sh:
```diff
@@ -5,14 +5,112 @@
 	return 0
 }
 
+platform_do_upgrade_combined() {
```
I don't really like the combined image format. In my opinion it is way to complex and depends on a lot of magic. For the [OpenMesh A62](0b204902075157dbb002ea04f41a5b1a4fbe848c) we managed to reuse parts of the nand upgrade code to write to the -not in a row - kernel and rootfs partition.

I'm not yet sure if we can do the same here, as I would expect that you would like to have a migration path from the ar71xx image to the ath79 image.

Before going into more details I would like your opinion about a new/different sysupgrade format.

tmn505 replied:
Thanks for providing another in-house solution, didn't know about this one. I'll have to look closely at this and experiment a bit, at first glance it looks pretty clean.

Personally don't have strong feelings about `combined-image`, I just went with old code which worked. Whichever solution is underneath doesn't matter for end user. What matters is swift upgrade, that's why I would like to keep the possibility to sysupgrade from ar71xx. Right now I can't say when I'll try to accommodate it, so I would like to keep combined-image for now.

mkresin commented on target/linux/ath79/image/generic-ubnt.mk:
```diff
@@ -117,3 +117,33 @@ define Device/ubnt_unifiac-pro
+  IMAGE/factory.bin := append-rootfs | pad-rootfs | mkubntimage | check-size $$$$(IMAGE_SIZE)
+  IMAGE/sysupgrade.bin := append-rootfs | pad-rootfs | combined-image | check-size $$$$(IMAGE_SIZE) | append-metadata
+  KERNEL := kernel-bin | append-dtb | lzma | pad-to $$(BLOCKSIZE)
+  KERNEL_INITRAMFS := kernel-bin | append-dtb | pad-to $$(BLOCKSIZE)
```
I'm quite sure the padding isn't required for the initramfs image

tofurky replied:
k

mkresin reviewed (changes requested):
Please use ```git commit --amend``` to add changes to your commit. Use ```git push --force``` to overwrite your github repository/branch. The PR will afterwards update automatically. Please don't close this PR and open a new one.

mkresin commented:
I don't really like the combined image format. In my opinion it is way to complex and depends on a lot of magic. For the [OpenMesh A62](0b204902075157dbb002ea04f41a5b1a4fbe848c) we managed to reuse parts of the nand upgrade code to write to the -not in a row - kernel and rootfs partition.

I'm not yet sure if we can do the same here, as I would expect that you would like to have a migration path from the ar71xx image to the ath79 image.

Before going into more details I would like your opinion about a new/different sysupgrade format.

tofurky commented on target/linux/ath79/image/generic-ubnt.mk:
```diff
@@ -117,3 +117,33 @@ define Device/ubnt_unifiac-pro
+  UBNT_BOARD := RSPRO
+  UBNT_TYPE := RSPRO
+  UBNT_CHIP := ar7100pro
+  DEVICE_PACKAGES += swconfig
```
the shared Device/ubnt_routerstation stanza sets `DEVICE_PACKAGES := kmod-usb-ohci kmod-usb2 fconfig`. in my testing this seemed to override `DEFAULT_PACKAGES` in target/linux/ath79/Makefile completely. would you rather have Device/ubnt_routerstation do `DEVICE_PACKAGES += -kmod-ath9k -wpad-mini -uboot-envtools kmod-usb-ohci kmod-usb2 fconfig`? taking that into consideration it seemed cleaner to do it this way. the only way i could get the DEVICE_PACKAGES it to influence the generated images was by setting `CONFIG_TARGET_PER_DEVICE_ROOTFS=y`. maybe i'm misunderstanding something?

mkresin replied:
DEFAULT_PACKAGES and DEVICE_PACKAGES are merged into the final package list and should work. Either we have a new bug or you did something wrong.

tofurky commented on target/linux/ath79/dts/ar7161_ubnt_rs.dtsi:
```diff
@@ -0,0 +1,114 @@
+	flash@0 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "mxicy,mx25l12805d", "jedec,spi-nor";
```
i previously had only `compatible = "mx25l12805d";` without the `mxicy,`, but it was suggested to add the full manufacturer name and also add the `"jdec,spi-nor"` as a secondary property. are you saying *no* compatible line is required? or that `jedec,spi-nor` is sufficient? please clarify, thanks.

tmn505 replied:
I prefer to fully describe the hardware, but documentation says 'May include a device-specific string consisting...', so I'm fine with dropping this.

tofurky commented on target/linux/ath79/dts/ar7161_ubnt_rs.dtsi:
```diff
@@ -0,0 +1,114 @@
+
+	gpio-keys-polled {
+		compatible = "gpio-keys-polled";
+		#address-cells = <1>;
```
i used the existing ath79 .dts as a reference here:
```
matt@aquos:~/devel/openwrt/openwrt$ fgrep -r -A3 'gpio-keys-polled {' target/linux/ath79/dts/
target/linux/ath79/dts/ar7161_netgear_wndr3800.dts:	gpio-keys-polled {
target/linux/ath79/dts/ar7161_netgear_wndr3800.dts-		compatible = "gpio-keys-polled";
target/linux/ath79/dts/ar7161_netgear_wndr3800.dts-		#address-cells = <1>;
target/linux/ath79/dts/ar7161_netgear_wndr3800.dts-		#size-cells = <0>;
--
target/linux/ath79/dts/ar7240_netgear_wnr612-v2.dtsi:	gpio-keys-polled {
target/linux/ath79/dts/ar7240_netgear_wnr612-v2.dtsi-		compatible = "gpio-keys-polled";
target/linux/ath79/dts/ar7240_netgear_wnr612-v2.dtsi-		#address-cells = <1>;
target/linux/ath79/dts/ar7240_netgear_wnr612-v2.dtsi-		#size-cells = <0>;
--
target/linux/ath79/dts/ar7240_tl-wr740n-v2.dts:	gpio-keys-polled {
target/linux/ath79/dts/ar7240_tl-wr740n-v2.dts-		compatible = "gpio-keys-polled";
target/linux/ath79/dts/ar7240_tl-wr740n-v2.dts-		#address-cells = <1>;
target/linux/ath79/dts/ar7240_tl-wr740n-v2.dts-		#size-cells = <0>;
--
target/linux/ath79/dts/ar7241_tp-link.dtsi:	gpio-keys-polled {
target/linux/ath79/dts/ar7241_tp-link.dtsi-		compatible = "gpio-keys-polled";
target/linux/ath79/dts/ar7241_tp-link.dtsi-		#address-cells = <1>;
target/linux/ath79/dts/ar7241_tp-link.dtsi-		#size-cells = <0>;
--
target/linux/ath79/dts/ar9331_tl-wr703n_tl-mr10u.dtsi:	gpio-keys-polled {
target/linux/ath79/dts/ar9331_tl-wr703n_tl-mr10u.dtsi-		compatible = "gpio-keys-polled";
target/linux/ath79/dts/ar9331_tl-wr703n_tl-mr10u.dtsi-		#address-cells = <1>;
target/linux/ath79/dts/ar9331_tl-wr703n_tl-mr10u.dtsi-		#size-cells = <0>;
--
target/linux/ath79/dts/ar9331_tl-wr741nd-v4.dtsi:	gpio-keys-polled {
target/linux/ath79/dts/ar9331_tl-wr741nd-v4.dtsi-		compatible = "gpio-keys-polled";
target/linux/ath79/dts/ar9331_tl-wr741nd-v4.dtsi-		#address-cells = <1>;
target/linux/ath79/dts/ar9331_tl-wr741nd-v4.dtsi-		#size-cells = <0>;
--
target/linux/ath79/dts/ar7161_ubnt_rs.dtsi:	gpio-keys-polled {
target/linux/ath79/dts/ar7161_ubnt_rs.dtsi-		compatible = "gpio-keys-polled";
target/linux/ath79/dts/ar7161_ubnt_rs.dtsi-		#address-cells = <1>;
target/linux/ath79/dts/ar7161_ubnt_rs.dtsi-		#size-cells = <0>;
```

tmn505 replied:
Agreed.

tofurky commented on target/linux/ath79/base-files/etc/board.d/02_network:
```diff
@@ -130,6 +130,14 @@ ath79_setup_interfaces()
+		ucidef_set_interfaces_lan_wan "eth1" "eth0"
+		;;
+	"ubnt,rspro")
+		ucidef_set_interfaces_lan_wan "eth1.1" "eth0"
```
k

tmn505 commented on target/linux/ath79/dts/ar7161_ubnt_rs.dts:
```diff
@@ -0,0 +1,27 @@
+#include "ar7161_ubnt_rs.dtsi"
+
+/ {
+	compatible = "ubnt,rs", "qca,ar7161";
```
That's fine with me.

tofurky replied:
i just did a build with the naming changes made, but it doesn't feel correct to me. ar71xx images are named like "lede-ar71xx-generic-ubnt-rspro-squashfs-factory.bin". ubiquiti themselves refer to the devices as "RS", "RSx", and "RSPRO" in their factory image metadata. we'd be using an abbreviated company name (ubnt) with an *abnormally long* board name. it really seems too long to me and a departure from how this board has historically been identified (and continues to be for ar71xx target). and, if only the dts compatible property and filenames were changed, then it becomes an inconsistent mess between the generic-ubnt.mk and dts files.
```
matt@aquos:~/devel/openwrt/openwrt$ xxd bin/targets/ath79/generic/openwrt-ath79-generic-ubnt_routerstationpro-squashfs-factory.bin|head
00000000: 4f50 454e 5253 5052 4f2e 6172 3731 3030  OPENRSPRO.ar7100
00000010: 7072 6f2e 7636 2e30 2e30 2d4f 7065 6e57  pro.v6.0.0-OpenW
00000020: 7274 2d72 3737 3739 2b31 2d38 3137 3438  rt-r7779+1-81748
00000030: 3533 6337 3800 0000 0000 0000 0000 0000  53c78...........
```

tmn505 replied:
Well, I do like the shorter name of the device, and keeping consistency with previous contributions/naming would be the best course of action, so I'll be in for keeping the short name. But if that's a road-blocker or undisputable I won't block the change to longer version.

tmn505 commented on target/linux/ath79/image/generic-ubnt.mk:
```diff
@@ -117,3 +117,33 @@ define Device/ubnt_unifiac-pro
+  KERNEL_INITRAMFS := kernel-bin | append-dtb | pad-to $$(BLOCKSIZE)
+endef
+
+define Device/ubnt_rs
```
That's fine with me.

adrianschmutzler commented on target/linux/ath79/base-files/etc/board.d/02_network:
```diff
@@ -130,6 +130,14 @@ ath79_setup_interfaces()
 		ucidef_add_switch "switch0" \
 			"0@eth0" "3:lan:1" "5:lan:2" "4:wan"
 		;;
+	"ubnt,rs")
```
Remove quotes here ...

adrianschmutzler commented on target/linux/ath79/base-files/etc/board.d/02_network:
```diff
@@ -130,6 +130,14 @@ ath79_setup_interfaces()
+	"ubnt,rs")
+		ucidef_set_interfaces_lan_wan "eth1" "eth0"
+		;;
+	"ubnt,rspro")
```
... and here. And maybe in the other switch-cases later in the commit.

tofurky commented:
A new version has been pushed with the following changes:

- rebase onto current master (there were a couple conflicts due to recent changes)
- remove the platform_check_combined function from lib/upgrade/platform.sh
- remove quotes from case statements
- remove size-cells/address-cells from gpio-keys-polled in .dtsi
- remove "mxicy,mx25l12805d" from flash compatible property in .dtsi
- remove padding step from initramfs generation
- change DEVICE_PACKAGES to explicitly exclude a few of the DEFAULT_PACKAGES as defined in target/linux/ath79/Makefile

I did take a look at how the lib/upgrade/platform.sh could be minimized. However, this limits it to just the rs/rspro and takes away the flexibility of supporting other boards (e.g. if they're migrated from ar71xx in the future), but it is not included in this update:
```
routerstation_do_upgrade() {
        CI_BLKSZ=65536
        local kern_length=0x$(dd if="$1" bs=2 skip=1 count=4 2>/dev/null)
        local kern_blocks=$(($kern_length / $CI_BLKSZ))
        local root_blocks=$((0x$(dd if="$1" bs=2 skip=5 count=4 2>/dev/null) / $CI_BLKSZ))

        if [ ${kern_blocks:-0} -gt 0 ] && [ ${root_blocks:-0} -gt 0 ]; then
                local append=""
                [ -f "$CONF_TAR" -a "$SAVE_CONFIG" -eq 1 ] && append="-j $CONF_TAR"
                ( dd if="$1" bs=$CI_BLKSZ skip=1 count=$kern_blocks 2>/dev/null; \
                  dd if="$1" bs=$CI_BLKSZ skip=$((1+$kern_blocks)) count=$root_blocks 2>/dev/null ) | \
                    mtd -r $append -Fkernel:$kern_length:0x80060000,rootfs write - 00010000:kernel:rootfs
        fi
}
```

mkresin commented:
> Well, I do like the shorter name of the device, and keeping consistency with previous contributions/naming would be the best course of action, so I'll be in for keeping the short name. But if that's a road-blocker or undisputable I won't block the change to longer version.

Okay, I'd asked three times to use the more descriptive userspace boardname and image name. It should be crystal clear that I would like to see this changed. Otherwise, I wouldn't demand the same change over and over again.

mkresin commented:
Regarding the routerstation_do_upgrade code. What is the purpose of dd'ing the kernel and rootfs in different steps? Can't we strip/skip the combined image header and write the result to the flash as it is?

I know that on ar71xx the kernel might be padded to an erase block boundary. We stopped doing it this way for some of the recently add ath79 images and it works just fine. Might be worse to give it a try.

If it's necessary to keep the erase block boundary padding, we should pad the kernel instead (which is already done and might not be required as well - see above).

tofurky commented:
Well, to be fair, you said it was your preference and not set in stone. That wasn't crystal clear.

I currently am still not convinced that the targets should be renamed, and I haven't seen any strong argument for doing so.

For this reason I am hesitant to sign off with my name on something that seems incorrect/unnecessary to me. Sorry, hopefully someone else can voice their opinion on this.

The reasons I see for keeping it are:

- Initial factory images from Ubiquiti for the boards were named like "RSx.ar7100.OpenWRT.bin" and "RSPRO.ar7100.OpenWRT.bin".
- The bootloader looks for 'RS' and 'RSPRO' in the headers of images when flashing via TFTP. This isn't something configurable in RedBoot.
- For most of the past decade, the images have been named like ubnt-rs-jffs2-sysupgrade.bin and ubnt-rspro-jffs2-factory.bin. This goes as far back as the boards' initial appearance in 10.03.
- Many other boards with long names, such as the Netgear Nighthawk X4S, have their images and devicetree referring to the Model #, e.g. r7800. rs/rspro are the closest things to model numbers these boards have.
- If we switched to the full names, the image filenames would be some of the longest in all of OpenWRT, e.g. openwrt-ath79-generic-ubnt_routerstationpro-squashfs-factory.bin.

mkresin commented:
I'm going to close this PR for now. At the moment the ath79 target changes to fast to keep PRs open for a longer time.

Please reopen the PR if you have the time to address the requested changes, rebase your PR to master and incorporated all the recently applied changes.

tofurky commented:
This has been rebased to current master (there were no conflicts though). All of the references to rs and rspro have been changed to routerstation and routerstation-pro.

mkresin commented:
Would you please add the minimised sysupgrade code as well. I'm seeing ath79 as a chance to do a lot of stuff different and better. Nobody really knows which board requires the overly complex combined sysupgrade code.

tofurky commented:
The "routerstation_do_upgrade" function has been added after some brief testing. I confirmed that it's enough to both sysupgrade from master to master, as well as from master back down to 17.01.5. The commit message has been updated to reflect this.

If I understand the role of SUPPORTED_DEVICES correctly, it might be helpful to add += ubnt-rs and ubnt-rspro so that sysupgrade -F isn't required from ar71xx, seeing as the target name changed. Many of the existing ath79 targets are already using it.

tofurky commented:
Actually nevermind, I did try adding the SUPPORTED_DEVICES strings but even though device name validation then passes, it still fails on checksum for some reason. So, I'll leave it as it is.

mkresin commented:
>> Regarding the routerstation_do_upgrade code. What is the purpose of dd'ing the kernel and rootfs in different steps? Can't we strip/skip the combined image header and write the result to the flash as it is?

>We need to extract kernel length from the image which is stored in combined image header. This value is provided later to `mtd` to update FIS partition table. For compatibility reasons with ar71xx the header should be kept, otherwise we force user to use factory image to flash it from bootloader or from running board, manually writing kernel and rootfs partitions, and updating the FIS partition to correct values.

I failed to point out which dd's I'm talking about. I'm talking about:

```
( dd if="$1" bs=$CI_BLKSZ skip=1 count=$kern_blocks 2>/dev/null; \
  dd if="$1" bs=$CI_BLKSZ skip=$((1+$kern_blocks)) count=$root_blocks 2>/dev/null ) | \
```

What I've in mind for a minimal version would look like:

```
# we're using the combined image format but can keep the sysupgrade simple
redboot_fis_do_upgrade() {
	local kern_length=0x$(dd if="$1" bs=2 skip=1 count=4 2>/dev/null)
	local append

	[ -f "$CONF_TAR" -a "$SAVE_CONFIG" -eq 1 ] && append="-j $CONF_TAR"
	dd if="$1" bs=64k skip=1 2>/dev/null | \
		mtd -r $append -Fkernel:$kern_length:0x80060000,rootfs write - 00010000:kernel:rootfs
}
```

But I guess I missed something and it isn't that easy.

> Actually nevermind, I did try adding the SUPPORTED_DEVICES strings but even though device name validation then passes, it still fails on checksum for some reason. So, I'll leave it as it is.

Is it the sysupgrade code that throws an checksum error? If so, it might be related to the appended image metadata and we would have a first user for the never working - and now removed - metadata stripping code (5e1b4c57ded7898be5255aef594fa18ec206f0b2).

>> I know that on ar71xx the kernel might be padded to an erase block boundary. We stopped doing it this way for some of the recently add ath79 images and it works just fine. Might be worse to give it a try.

> AFAIK RedBoot expects any partition to start at erase block boundary.

But does redboot need to ever touch our not EB aligned rootfs partition? But that is really the most unimportant detail.

> Padding rootfs can probably be dropped, I didn't experiment with that yet, I'll try it out today.

Nope, that is required for sure. I'm talking about the kernel padding to align the rootfs to an EB 

> I prepared another minimalistic approach to the sysupgrade, it drops the combined image header and uses default upgrade command.

I'm sorry to say, but it doesn't really look better than what we have at the moment. The search for the squashfs magic doesn't seem to me a good idea. Way to error prone for false positives and not really required if you know the kernel size during image creation.

> This is tested but breaks compatibility with ar71xx and would require user to use factory image or manually flashing partitions like I mentioned earlier

If breaking, than properly :-).

I strongly suggest to use the nand sysupgrade-tar format instead. By doing so, you'll have the kernel and rootfs as individual binaries. If you pipe the kernel to `wc -c`, you will get the kernel size. And that is what the whole redboot_fis sysupgrade code is about. No need to take care of unexpected dd output, no need to know the custom header length and so on.

I've the following code from an old gist as an example:

```
redboot_fis_do_upgrade() {
	local board_dir=$(tar tf $tar_file | grep -m 1 '^sysupgrade-.*/$')
	local kern_length=$(tar xf $tar_file ${board_dir}/kernel -O | wc -c)
	local append

	[ "$SAVE_CONFIG" -eq 1 ] && append="${MTD_CONFIG_ARGS} -j ${CONF_TAR}"
	tar xf $tar_file ${board_dir}/kernel ${board_dir}/rootfs -O | \
		mtd -r $append -Fkernel:$kern_length:0x80060000,rootfs write - 00010000:kernel:rootfs
}
```

The image build code would need to be similar to https://github.com/openwrt/openwrt/blob/6b4ba118ac1ba1318182eb3a22864e86165f0b09/target/linux/ipq40xx/image/Makefile#L153

Nevertheless, breaking compatibility with ar71xx due to a different image format, is something that should be discussed on the OpenWrt mailinglist. It's I topic I don't feel save to make any decisions without a public discussion in advance.

tmn505 commented:
> What I've in mind for a minimal version would look like:
````
# we're using the combined image format but can keep the sysupgrade simple
redboot_fis_do_upgrade() {
	local kern_length=0x$(dd if="$1" bs=2 skip=1 count=4 2>/dev/null)
	local append

	[ -f "$CONF_TAR" -a "$SAVE_CONFIG" -eq 1 ] && append="-j $CONF_TAR"
	dd if="$1" bs=64k skip=1 2>/dev/null | \
		mtd -r $append -Fkernel:$kern_length:0x80060000,rootfs write - 00010000:kernel:rootfs
}
````
> But I guess I missed something and it isn't that easy.

Don't know why someone decided to pipe each partition when creating this part, as mentioned earlier I did simply copy most of old code. Your approach also works, so this can can be simplified as in Your example, minus erase size `00010000`, `mtd` interprets this as partition name. 

>> Actually nevermind, I did try adding the SUPPORTED_DEVICES strings but even though device name validation then passes, it still fails on checksum for some reason. So, I'll leave it as it is.

> Is it the sysupgrade code that throws an checksum error? If so, it might be related to the appended image metadata and we would have a first user for the never working - and now removed - metadata stripping code (5e1b4c5).

As I vaguely remember I had the same error about two months ago so this commit should be unrelated, but to make it sure I'll have to test it with this commit reverted.

> But does redboot need to ever touch our not EB aligned rootfs partition? But that is really the most unimportant detail.

I'm not confident about it, will have to check.

> Nope, that is required for sure. I'm talking about the kernel padding to align the rootfs to an EB

Yep.

> I'm sorry to say, but it doesn't really look better than what we have at the moment. The search for the squashfs magic doesn't seem to me a good idea. Way to error prone for false positives and not really required if you know the kernel size during image creation.

Yes, I also didn't like the string search. I had two ideas: the first was this one checking kernel length on target, and second one was to embed kernel length in metadata on the process of creating image. Went with faster implementation approach. Would metadata be good place for this?

>> This is tested but breaks compatibility with ar71xx and would require user to use factory image or manually flashing partitions like I mentioned earlier

> If breaking, than properly :-).

Agreed.

> I strongly suggest to use the nand sysupgrade-tar format instead. By doing so, you'll have the kernel and rootfs as individual binaries. If you pipe the kernel to wc -c, you will get the kernel size. And that is what the whole redboot_fis sysupgrade code is about. No need to take care of unexpected dd output, no need to know the custom header length and so on.

> I've the following code from an old gist as an example:

````
redboot_fis_do_upgrade() {
	local board_dir=$(tar tf $tar_file | grep -m 1 '^sysupgrade-.*/$')
	local kern_length=$(tar xf $tar_file ${board_dir}/kernel -O | wc -c)
	local append

	[ "$SAVE_CONFIG" -eq 1 ] && append="${MTD_CONFIG_ARGS} -j ${CONF_TAR}"
	tar xf $tar_file ${board_dir}/kernel ${board_dir}/rootfs -O | \
		mtd -r $append -Fkernel:$kern_length:0x80060000,rootfs write - 00010000:kernel:rootfs
}
````
[...]

TBF I didn't look at it properly yet, will do in upcoming days.

> Nevertheless, breaking compatibility with ar71xx due to a different image format, is something that should be discussed on the OpenWrt mailinglist. It's I topic I don't feel save to make any decisions without a public discussion in advance.

Acknowledged.

mkresin commented:
>> Is it the sysupgrade code that throws an checksum error? If so, it might be related to the appended image metadata and we would have a first user for the never working - and now removed - metadata stripping code (5e1b4c5).

> As I vaguely remember I had the same error about two months ago so this commit should be unrelated, but to make it sure I'll have to test it with this commit reverted.

As written, the image metadata stripping never worked. Reverting the commit is not going to change anything. it will just restore not wokring code. please build an ath79 image without append-metadata and test again.

> Yes, I also didn't like the string search. I had two ideas: the first was this one checking kernel length on target, and second one was to embed kernel length in metadata on the process of creating image. Went with faster implementation approach. Would metadata be good place for this?

From my current understanding of the metadata it would be the wrong place. beside the fact that no such code exists. getting the kernel size from the combined image header or via `wc -c` seem sufficient to me.

tofurky commented:
The following does work for master -> master and master -> 17.01.x:
```
routerstation_do_upgrade() {
	local kern_length=0x$(dd if="$1" bs=2 skip=1 count=4 2>/dev/null)

	[ $(printf '%d' "${kern_length:-0}") -gt 0 ] || { echo "Bad length, aborting!"; return; }

	local append=""
	[ -f "$CONF_TAR" -a "$SAVE_CONFIG" -eq 1 ] && append="-j $CONF_TAR"
	dd if="$1" bs=64k skip=1 2>/dev/null | \
		mtd -r $append -Fkernel:$kern_length:0x80060000,rootfs write - kernel:rootfs
}
```
Since all the other checks were removed I added one back in for $kern_length. I'm a bit hesistant to just blindly pass the result of some shell math to mtd. Maybe there is a more thorough way?

Regarding modifying the sysupgrade mechanism - that's fine, I don't think this PR should be the place for it though. It can always be added in later. I would like to see this merged and I can see that discussion adding weeks onto the timeline.

mkresin commented:
None of the base sysupgrade scripts are doing these kind of sanity check and we should follow this approach. In fact, with the switch to forced metadata validation, this kind of code was removed from the targets.

Can we rename the function to "redboot_fis_do_upgrade"? It doesn't need to be a routerstation specific upgrade function in future. What we are really dealing with is redboot+fis. 

Please move the `local append=""` to the top. Our script style is define variables at the beginning of a function. You can safely change it to `local append`. It's empty in both cases and the version without adding an empty string is what we're using in our scripts.

tofurky commented:
From what I can see, the fwtool image validation added in cc853810a4aae04f1afcad3439bf9db6fc1f4852 does not verify that the header is correct. It seems to focus on the json metadata at the end like so:
`{  "supported_devices":["ubnt,routerstation-pro"], "version": { "dist": "OpenWrt", "version": "SNAPSHOT", "revision": "r7904+1-6b4ba118ac", "board": "ath79" } }`

Then there is the checksum for the actual kernel/squashfs content which  *after* the size header:
```
00000000: 4349 3030 3139 3030 3030 3030 3230 3030  CI00190000002000
00000010: 3030 3432 3634 3937 3066 6366 3866 3265  004264970fcf8f2e
00000020: 3436 6635 6337 6135 6336 3066 6237 6461  46f5c7a5c60fb7da
00000030: 3537 0000 0000 0000 0000 0000 0000 0000  57..............
```

Before making this change, I also skimmed the source of mtd.c/fis.c to see if it performed any sanity checks on the arguments to the new FIS table passed to `-F`, but it doesn't appear to.

Maybe the validation of the size header is happening somewhere I missed?

Perhaps ucert image check could catch this, but not everyone will use this.

This being said, do we still blindly want write the unchecked result of `kern_length` to the FIS table? It gives me pause as someone who uses this board as their primary router/AP.

About the rename, sure, if you want to. I'm not certain if the function in its current state would be suitable for any other boards besides the RouterStation(Pro), though. I did some searching to see which other boards use RedBoot and there are a few ar71xx like https://openwrt.org/toh/ruckus/zf7962 and https://wiki.openwrt.org/toh/evaluation.boards/pb44 which seem to have non-identical partition layouts.

tofurky commented:
The `local append` has been moved to the top.

Rebased on master as of now, no conflicts and no other changes e.g. LED handling needed to be accounted for.

mkresin commented:
> From what I can see, the fwtool image validation added in cc85381 does not verify that the header is correct. It seems to focus on the json metadata at the end like so:

I'm not sure if I can follow, respectively I've no idea what you are looking for.

The size field of the CI image is added by the image build code and matches the size of the kernel/roots passed to the combined-image image build recipe. Adding the same size field to the image metadata wouldn't really change anything as it would have the same value.

The image metadata have a checksum. Not sure yet if the checksum covers only the metadata or the whole image. For the later, it would be safe to assume that an image transferred to the device doesn't have bitflips if the checksum can be validated.

> This being said, do we still blindly want write the unchecked result of kern_length to the FIS table? It gives me pause as someone who uses this board as their primary router/AP.

Yes this is what I asked to change. As written before, we are doing it this way for the common sysupgrade code as well. We neither validate the size field of the uImage header nor the size of any binaries. If you include a zero byte kernel in the NAND sysupgrade tar archive, it will brick the device as well. The general idea is to ensure during image creation that everything is fine (see check-size for example). If some sanity checks are missing, they should be added to the image build code.

The trend is rather to get rid of on device validation as pointed out above. Doing it on the device will result in a lot of hard to maintain sysupgrade code, similar to what we have in ar71xx right now.

If you feel the urge of doing on device image validation, it should implemented for the common sysupgrade code first. Such a proposal should be send for discussion to the mailing list to get all pros and cons of adding such code.

rmilecki commented on target/linux/ath79/dts/ar7161_ubnt_routerstation.dtsi:
```diff
@@ -0,0 +1,119 @@
+			compatible = "ecoscentric,redboot-fis-partitions";
+		};
+
+		rbc: partition@fff000 {
```
Yeah, that way of mixing old syntax with new one to trigger two parsers (`ecoscentric,redboot-fis-partitions` and `fixed-partitions`) seems really hacky to me. Don't do that as you're trying to describe the same thing in two conflicting ways at the same time.

So it has to be either: `ecoscentric,redboot-fis-partitions` or `fixed-partitions`.

I guess `ecoscentric,redboot-fis-partitions` is needed because flash layout isn't fixed. So go with that and then either:
1. Use user-space (as suggested by @mkresin) to read MAC
2. OR Rewrite `redboot-fis-partitions` to allow specifying nodes

The first solution should be much easier.

tofurky commented:
MAC addresses are now being set in userspace and the mtd-mac-address/fixed-partions has been removed from the .dtsi. Commit message is adjusted to reflect the necessity of fconfig and explain why things were done this way.

tofurky committed:
ath79: initial support for Ubiquiti RouterStation and RouterStation Pro

This adds a shared ar7161_ubnt_routerstation.dtsi as well as two other
.dts files that utilize it, ar7161_ubnt_routerstation.dts and
ar7161_ubnt_routerstation-pro.dts.

The modifications to generic-ubnt.mk, config-default, and base-files
necessary for image generation, parsing RedBoot FIS partitions, network
configuration, and sysupgrade are also included.

This reintroduces vital bits from platform_do_upgrade_combined() and its
supporting functions to /lib/upgrade/platform.sh, which were previously
removed from ath79 in 3e9d9f62258f80298710441e0db557e59e152dcf "ath79:
sysupgrade: drop unused platform checks". The new function is called
"routerstation_do_upgrade" and will *only* work for the RouterStation
series of boards. It does however retain the ability to downgrade (e.g.
from master -> 17.01.x using sysupgrade -F).

All hardware is functional including the AR8216 switch (for the Pro),
wireless via ath5k/ath9k using the miniPCI slots, flash, USB, button,
and LED.

Switch and LAN/WAN configuration is the same as it is with the
equivalent ar71xx targets. MAC addresses are assigned based upon the
content stored in the RedBoot config partition.

Flashing via both sysupgrade and TFTP has been confirmed to work. Also,
the initramfs images are now raw .bin files instead of being wrapped in
a uImage (as they currently are in ar71xx), which makes them bootable
with RedBoot.

One notable difference to ar71xx is the inclusion of the RedBoot
"fconfig" utility (analogous to U-Boot’s fw_printenv/fw_setenv) in
DEVICE_PACKAGES. The FIS partitions are probed using the RedBoot MTD
parser’s DT binding, whose proper usage is mutually exclusive to
defining a separate fixed-partitions node for "RedBoot config". This
config partition contains the board's base MAC address. The lack of a
hard-coded flash location means that the mtd-mac-address property cannot
be used in the .dts, so instead fconfig is used to read the MAC
addresses from flash in userspace during first boot.

Signed-off-by: Tomasz Maciej Nowak <tomek_n@o2.pl>
Signed-off-by: Matt Merhar <mattmerhar@protonmail.com>

mkresin commented:
merged, thanks!
